### PR TITLE
Multi sdram

### PR DESCRIPTION
**Pull request by Christian-B**
Builds on https://github.com/SpiNNakerManchester/sPyNNaker/pull/803

Includes one extra fix of a few bytes to many in the recording regions.

Depends on https://github.com/SpiNNakerManchester/PACMAN/pull/297  as it uses the WithMallocSdram
Other vertexes not yet done with MultiRegionSDRAM

This opens the way for the SynapticManager to use variableSDRAM costs to reduce the guess work but that was considered out of scope for this PR.

tested with https://github.com/SpiNNakerManchester/sPyNNaker8/pull/380

### CODE REVIEW

File: spynnaker/pyNN/models/common/neuron_recorder.py [hunk 1]
```diff
@@ -654,7 +654,8 @@ def get_sdram_usage_in_bytes(self, vertex_slice):
         return ((self.N_ITEM_TYPES * DataType.UINT32.size) + var_bytes +
                 bitfield_bytes)
 
-    def _get_fixed_sdram_usage(self, vertex_slice):
+    def _get_indexes_sdram_usage(self, vertex_slice):
+        """ DEAD I think"""
```
Comment (rowleya) [inline spynnaker/pyNN/models/common/neuron_recorder.py]: This was in use in master, but maybe it went away?

File: spynnaker/pyNN/models/common/neuron_recorder.py [hunk 2]
```diff
@@ -666,34 +667,44 @@ def _get_fixed_sdram_usage(self, vertex_slice):
...
+        Find the SDRAM cost for recording this slice
+
+        :param vertex_slice: Slice being costed
+        :return: SDRAM cost
+        rtype: MultiRegionSDRAM
+        """
+        costs = MultiRegionSDRAM()
         for variable in self.__sampling_rates:
             rate = self.__sampling_rates[variable]
-            fixed_sdram += self._get_fixed_sdram_usage(vertex_slice)
```
Comment (rowleya) [inline spynnaker/pyNN/models/common/neuron_recorder.py]: This is where the method now marked as dead was called; thus it is this change that makes the code dead!

File: spynnaker/pyNN/models/neuron/abstract_population_vertex.py [hunk 3]
```diff
@@ -262,17 +264,32 @@ def _get_sdram_usage_for_neuron_params(self, vertex_slice):
...
-            SYSTEM_BYTES_REQUIREMENT +
-            self._get_sdram_usage_for_neuron_params(vertex_slice) +
-            self._neuron_recorder.get_static_sdram_usage(vertex_slice) +
+        costs = MultiRegionSDRAM()
+        region_costs = MultiRegionSDRAM()
+        # These values can be cross checked with the data_spec_text_files
+        region_costs.add_cost(
+            POPULATION_BASED_REGIONS.SYSTEM, SIMULATION_N_BYTES)
+        region_costs.add_cost(
+            POPULATION_BASED_REGIONS.NEURON_RECORDING,
```
Comment (alan-stokes) [inline spynnaker/pyNN/models/neuron/abstract_population_vertex.py:273]: wrong dsg region.

File: spynnaker/pyNN/models/neuron/abstract_population_vertex.py [hunk 4]
```diff
@@ -262,17 +264,32 @@ def _get_sdram_usage_for_neuron_params(self, vertex_slice):
...
+                vertex_slice, machine_time_step, graph, self))
+        region_costs.add_cost(
+            POPULATION_BASED_REGIONS.PROFILING,
             profile_utils.get_profile_region_size(self.__n_profile_samples))
-
-        return sdram_requirement
+        region_costs.add_cost(
+            "dsg overhead",
+            APP_PTR_TABLE_HEADER_BYTE_SIZE + MAX_MEM_REGIONS * 4)
+        costs.nest("Data_Spec_size", region_costs)
```
Comment (alan-stokes) [inline spynnaker/pyNN/models/neuron/abstract_population_vertex.py:290]: named constant please

File: spynnaker/pyNN/models/neuron/abstract_population_vertex.py [hunk 5]
```diff
@@ -262,17 +264,32 @@ def _get_sdram_usage_for_neuron_params(self, vertex_slice):
...
+        costs.add_cost("DSG_malloc", SARK_PER_MALLOC_SDRAM_USAGE)
```
Comment (alan-stokes) [inline spynnaker/pyNN/models/neuron/abstract_population_vertex.py:291]: named constant please

**Comment by Christian-B** [comment]
This is so old and so out of date not worth keeping

**Comment by rowleya** [comment]
Even better, we already use the MultiRegionSDRAM from elsewhere :).

**Pull request closed without merging**